File: iOSDevTips/Classes/CircularLoaderView/CircularImageView.m
//
//  CircularImageView.m
//  iOSDevTips
//
//  Created by piglikeyoung on 15/9/2.
//  Copyright (c) 2015年 piglikeyoung. All rights reserved.
//

#import "CircularImageView.h"
#import "CircularLoaderView.h"
#import "UIImageView+WebCache.h"

@interface CircularImageView()

@property (nonatomic, weak) CircularLoaderView *progressIndicatorView;

@end

@implementation CircularImageView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setUpLoaderView];
    }
    return self;
}

- (void) setUpLoaderView {
    // 创建加载的View
    CircularLoaderView *progressIndicatorView = [[CircularLoaderView alloc] initWithFrame:CGRectZero];
    self.progressIndicatorView = progressIndicatorView;
    [self addSubview:progressIndicatorView];
}


- (void)configureImageViewWithImageURL:(NSString *)sUrl animated:(BOOL)animated {
    
    NSURL *url = [NSURL URLWithString:sUrl];
    if (animated) {
        self.progressIndicatorView.frame = self.bounds;
        [self.progressIndicatorView setAutoresizingMask:UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight];
        // 下载图片
        [self sd_setImageWithURL:url placeholderImage:nil options:SDWebImageRetryFailed progress:^(NSInteger receivedSize, NSInteger expectedSize) {
            // 更新进度条
            self.progressIndicatorView.progress = @(receivedSize).floatValue / @(expectedSize).floatValue;
        } completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, NSURL *imageURL) {
            // 加载完成reveal动画
            [self.progressIndicatorView reveal];
        }];
    } else {
        self.progressIndicatorView.frame = CGRectZero;
        [self sd_setImageWithURL:url];
    }
}
@end

File: iOSDevTips/Classes/CircularLoaderView/Swift/CircularImageView.swift
//
//  CircularImageView.swift
//  iOSDevTips
//
//  Created by piglikeyoung on 15/9/2.
//  Copyright (c) 2015年 piglikeyoung. All rights reserved.
//

import UIKit

class CircularImageView: UIImageView {

    let progressIndicatorView = CircularLoaderView(frame: CGRectZero)
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(self.progressIndicatorView)
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func configureImageViewWithImageURL(sUrl : NSString, animated : Bool) {
        
        let url = NSURL(string: sUrl as String)
        if animated {
            progressIndicatorView.frame = bounds
            progressIndicatorView.autoresizingMask = .FlexibleWidth | .FlexibleHeight
            
            // 下载图片
            self.sd_setImageWithURL(url, placeholderImage: nil, options: .RetryFailed , progress: { [weak self](receivedSize, expectedSize) -> Void in
                self!.progressIndicatorView.progress = CGFloat(receivedSize)/CGFloat(expectedSize)
                }) { [weak self](image, error, _, _) -> Void in
                    self!.progressIndicatorView.reveal()
            }
        } else {
            progressIndicatorView.frame = CGRectZero
            self.sd_setImageWithURL(url)
        }
    }
}
